src/compiler/rust.rs

```diff
@@ -1184,6 +1427,239 @@ impl pkg::ToolchainPackager for RustToolchainPackager {
+    {
+        use std::mem;
+
+        // TODO: unfortunately there is exactly nothing you can do with the k given the
+        // current trait bounds. Just use some kind of sane value;
+        //let k_size = mem::size_of::<PathBuf>() + k.capacity();
+        let k_size = 3*8 + 100;
+
+        let crate_names_size: usize = v.deps.iter().map(|s| s.capacity()).sum();
+        let v_size: usize =
+            mem::size_of::<RlibDepsDetail>() + // Systemtime and vec itself
+            v.deps.capacity() * mem::size_of::<String>() + // Each string in the vec
+            crate_names_size; // Contents of all strings
+
+        k_size + v_size
+    }
+}
+
+#[cfg(feature = "dist-client")]
+#[derive(Debug)]
+struct RlibDepReader {
+    cache: Mutex<lru_cache::LruCache<PathBuf, RlibDepsDetail, RandomState, DepsSize>>,
+    executable: PathBuf,
+}
+
+#[cfg(feature = "dist-client")]
+impl RlibDepReader {
+    fn new_with_check(executable: PathBuf, env_vars: &[(OsString, OsString)]) -> Result<Self> {
+        use tempfile;
+
+        let temp_dir = tempfile::tempdir()
+            .chain_err(|| "Could not create temporary file for rlib output")?;
+        let temp_rlib = temp_dir.path().join("x.rlib");
+
+        let mut cmd = process::Command::new(&executable);
+        cmd
+            .arg("--crate-type=rlib")
+            .arg("-o").arg(&temp_rlib)
+            .arg("-")
+            .env_clear()
+            .envs(ref_env(env_vars));
+
+        let process::Output { status, stdout, stderr } = cmd.output()?;
+
+        if !status.success() {
+            bail!("Failed to compile a minimal rlib with {}", executable.display())
+        }
+        if !stdout.is_empty() {
+            bail!("rustc stdout non-empty when compiling a minimal rlib: {:?}", String::from_utf8_lossy(&stdout))
+        }
+        if !stderr.is_empty() {
+            bail!("rustc stderr non-empty when compiling a minimal rlib: {:?}", String::from_utf8_lossy(&stderr))
+        }
+
+        // The goal of this cache is to avoid repeated lookups when building a single project. Let's budget 3MB.
+        // Allowing for a 100 byte path, 50 dependecies per rlib and 20 characters per crate name, this roughly
+        // approximates to `path_size + path + vec_size + num_deps * (systemtime_size + string_size + crate_name_len)`
+        //                 `   3*8    +  100 +   3*8    +    50    * (      8         +     3*8     +       20      )`
+        //                 `2748` bytes per crate
+        // Allowing for possible overhead of up to double (for unused space in allocated memory), this means we
+        // can cache information from about 570 rlibs - easily enough for a single project.
+        const CACHE_SIZE: u64 = 3*1024*1024;
+        let cache = lru_cache::LruCache::with_meter(CACHE_SIZE, DepsSize);
+
+        let rlib_dep_reader = RlibDepReader { cache: Mutex::new(cache), executable };
+        if let Err(e) = rlib_dep_reader.discover_rlib_deps(env_vars, &temp_rlib) {
```

Comment:
Is this just a sanity check to ensure that `-Z ls` works at all with the current rustc?